File: me_daily/2025/11/02/cf2153d/cf2153d.py
from collections import defaultdict
from typing import List
import sys
from collections import Counter
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()

def solve():
    n = int(input())
    a = list(map(int, input().split()))
    
    def cost2(x, y):
        # 使两个元素相等的最小代价
        return abs(x - y)
    
    def cost3(x, y, z):
        # 使三个元素相等的最小代价（使用中位数）
        vals = sorted([x, y, z])
        median = vals[1]
        return abs(vals[0] - median) + abs(vals[2] - median)
    
    def solve_linear(arr):
        # 解决线性数组的DP
        m = len(arr)
        
        dp = [float('inf')] * (m + 1)
        dp[0] = 0
        
        for i in range(2, m + 1):
            # 尝试长度为2的块
            dp[i] = min(dp[i], dp[i-2] + cost2(arr[i-2], arr[i-1]))
            
            # 尝试长度为3的块
            if i >= 3:
                dp[i] = min(dp[i], dp[i-3] + cost3(arr[i-3], arr[i-2], arr[i-1]))
        
        return dp[m]
    
    ans = float('inf')
    
    # 尝试3个可能的断点
    for cut in range(3):
        if cut >= n:
            continue
        
        # 重新排列数组：从cut位置开始
        reordered = []
        for i in range(n):
            reordered.append(a[(cut + i) % n])
        
        # 计算线性数组的最优解
        result = solve_linear(reordered)
        ans = min(ans, result)
    
    print(ans)

t = II()
for _ in range(t):
    solve()

File: me_daily/2025/11/02/cf2153d/cf2153d.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func cost2(x, y int64) int64 {
	return abs(x - y)
}

func cost3(x, y, z int64) int64 {
	a := []int64{x, y, z}
	sort.Slice(a, func(i, j int) bool { return a[i] < a[j] })
	return abs(a[1]-a[0]) + abs(a[2]-a[1])
}

func solve(a []int64) int64 {
	n := len(a)
	const INF int64 = 1e12
	ans := INF

	// 尝试3个可能的断点
	for cut := 0; cut < 3; cut++ {
		if cut >= n {
			continue
		}

		// 重新排列数组
		re := make([]int64, n)
		for i := 0; i < n; i++ {
			re[i] = a[(i+cut)%n]
		}

		// DP求解
		f := make([]int64, n+1)
		for i := range f {
			f[i] = INF
		}
		f[0] = 0

		for i := 2; i <= n; i++ {
			// 长度为2的块
			f[i] = min(f[i], f[i-2]+cost2(re[i-2], re[i-1]))
			// 长度为3的块
			if i >= 3 {
				f[i] = min(f[i], f[i-3]+cost3(re[i-3], re[i-2], re[i-1]))
			}
		}

		ans = min(ans, f[n])
	}

	return ans
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	t, _ := strconv.Atoi(scanner.Text())

	for i := 0; i < t; i++ {
		scanner.Scan()
		n, _ := strconv.Atoi(scanner.Text())

		scanner.Scan()
		line := scanner.Text()
		parts := strings.Fields(line)
		a := make([]int64, n)
		for j := 0; j < n; j++ {
			a[j], _ = strconv.ParseInt(parts[j], 10, 64)
		}

		fmt.Println(solve(a))
	}
}
